File: apps/api/src/utils/session.rs
/// セッション管理ユーティリティ

use jsonwebtoken::{encode, decode, DecodingKey, EncodingKey, Header, Validation, Algorithm};
use serde::{Deserialize, Serialize};
use chrono::{Duration, Utc};
use crate::error::{ApiError, error_codes};

/// セッションJWTペイロード
#[derive(Debug, Serialize, Deserialize)]
struct SessionClaims {
    sub: String, // user_id
    exp: i64,    // 有効期限
    iat: i64,    // 発行時刻
}

/// セッションの有効期限（7日間）
const SESSION_DURATION_DAYS: i64 = 7;

/// セッションJWTを生成
pub fn generate_session_token(user_id: &str, secret: &str) -> Result<String, ApiError> {
    let now = Utc::now();
    let exp = now + Duration::days(SESSION_DURATION_DAYS);

    let claims = SessionClaims {
        sub: user_id.to_string(),
        exp: exp.timestamp(),
        iat: now.timestamp(),
    };

    let header = Header::default();
    let encoding_key = EncodingKey::from_secret(secret.as_bytes());

    encode(&header, &claims, &encoding_key)
        .map_err(|e| {
            tracing::error!("Failed to encode session token: {}", e);
            ApiError::InternalError(format!("Failed to generate session token: {}", e))
        })
}

/// セッションJWTを検証してユーザーIDを取得
#[allow(dead_code)]
pub fn verify_session_token(token: &str, secret: &str) -> Result<String, ApiError> {
    let validation = Validation::new(Algorithm::HS256);
    let decoding_key = DecodingKey::from_secret(secret.as_bytes());

    let token_data = decode::<SessionClaims>(token, &decoding_key, &validation)
        .map_err(|e| {
            tracing::error!("Failed to verify session token: {}", e);
            ApiError::Unauthorized {
                code: Some(error_codes::INVALID_SESSION.to_string()),
            }
        })?;

    Ok(token_data.claims.sub)
}

File: apps/web1/src/lib/api/lib/session.ts
// セッション管理ユーティリティ

import { SignJWT, jwtVerify } from "jose";
import { ERROR_CODES } from "./error-codes";

// セッションJWTペイロード
interface SessionClaims {
  sub: string; // user_id
  exp?: number; // 有効期限
  iat?: number; // 発行時刻
}

// セッションの有効期限（7日間）
const SESSION_DURATION_DAYS = 7;

/**
 * セッションJWTを生成
 */
export async function generateSessionToken(
  userId: string,
  secret: string
): Promise<string> {
  const now = Math.floor(Date.now() / 1000);
  const exp = now + SESSION_DURATION_DAYS * 24 * 60 * 60; // 7日後

  const token = await new SignJWT({ sub: userId })
    .setProtectedHeader({ alg: "HS256" })
    .setExpirationTime(exp)
    .setIssuedAt(now)
    .sign(new TextEncoder().encode(secret));

  return token;
}

/**
 * セッションJWTを検証してユーザーIDを取得
 */
export async function verifySessionToken(
  token: string,
  secret: string
): Promise<string> {
  try {
    const { payload } = await jwtVerify<SessionClaims>(
      token,
      new TextEncoder().encode(secret)
    );

    if (!payload.sub) {
      throw new Error("Invalid session token: missing sub claim");
    }

    return payload.sub;
  } catch (error) {
    console.error("Failed to verify session token:", error);
    throw new SessionError("Invalid session token", ERROR_CODES.INVALID_SESSION);
  }
}

// エラークラス
export class SessionError extends Error {
  constructor(
    message: string,
    public code: string = ERROR_CODES.INVALID_SESSION
  ) {
    super(message);
    this.name = "SessionError";
  }
}
